brave-instrumentation/aws-java-sdk-sqs/src/main/java/brave/instrumentation/aws/sqs/SendMessageTracingRequestHandler.java

```diff
@@ -0,0 +1,131 @@
+        @Override
+        public void put(Map<String, MessageAttributeValue> carrier, String key, String value) {
+          carrier.put(key,
+              new MessageAttributeValue().withDataType("String").withStringValue(value));
+        }
+      };
+
+  static final Propagation.Getter<Map<String, MessageAttributeValue>, String> GETTER =
+      new Propagation.Getter<Map<String, MessageAttributeValue>, String>() {
+        @Override public String get(Map<String, MessageAttributeValue> carrier, String key) {
+          return carrier.containsKey(key) ? carrier.get(key).getStringValue() : null;
+        }
+      };
+
+  final Tracing tracing;
+  final Tracer tracer;
+  final CurrentTraceContext currentTraceContext;
+  final TraceContext.Injector<Map<String, MessageAttributeValue>> injector;
+  final TraceContext.Extractor<Map<String, MessageAttributeValue>> extractor;
+
+  SendMessageTracingRequestHandler(Tracing tracing) {
+    this.tracing = tracing;
+    this.tracer = tracing.tracer();
+    this.currentTraceContext = tracing.currentTraceContext();
+    this.injector = tracing.propagation().injector(SETTER);
+    this.extractor = tracing.propagation().extractor(GETTER);
+  }
+
+  @Override
+  public AmazonWebServiceRequest beforeExecution(AmazonWebServiceRequest request) {
+    if (request instanceof SendMessageRequest) {
+      handleSendMessageRequest((SendMessageRequest) request);
+    } else if (request instanceof SendMessageBatchRequest) {
+      handleSendMessageBatchRequest((SendMessageBatchRequest) request);
+    }
+    return request;
+  }
+
+  private void handleSendMessageRequest(SendMessageRequest request) {
+    injectPerMessage(request.getQueueUrl(), request.getMessageAttributes());
+  }
+
+  private void handleSendMessageBatchRequest(SendMessageBatchRequest request) {
+    TraceContext maybeParent = currentTraceContext.get();
+
+    Span span;
+    if (maybeParent == null) {
+      span = tracer.nextSpan();
+    } else {
+      // If we have a span in scope assume headers were cleared before
+      span = tracer.newChild(maybeParent);
+    }
+
+    span.name("publish-batch").start();
+    try (Tracer.SpanInScope scope = tracer.withSpanInScope(span)) {
+      for (SendMessageBatchRequestEntry entry : request.getEntries()) {
+        injectPerMessage(request.getQueueUrl(), entry.getMessageAttributes());
+      }
+    } finally {
+      span.finish();
+    }
+  }
+
+  private void injectPerMessage(String queueUrl,
+      Map<String, MessageAttributeValue> messageAttributes) {
+    TraceContext maybeParent = currentTraceContext.get();
+
+    Span span;
+    if (maybeParent == null) {
+      span = tracer.nextSpan(extractAndClearHeaders(messageAttributes));
+    } else {
+      // If we have a span in scope assume headers were cleared before
+      span = tracer.newChild(maybeParent);
+    }
+
+    if (!span.isNoop()) {
+      span.kind(Span.Kind.PRODUCER).name("publish");
+      span.remoteServiceName("amazon-sqs/" + queueUrl);
```

Comment:
I don't think we want urls in service names as it will blow up the index.

Reply:
Got it, moved to tag